OpenSSL output formats and interoperability draft

mgerstner opened the pull request:
Follow-up on PR #427. I don't expect this to be accepted yet, I just would like to know if this matches your expectations, before I start touching a dozen of other tools and writing documentation for it.

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+#include "files.h"
+#include "tpm2_util.h"
+#include "tpm2_alg_util.h"
+#include "conversion.h"
```
sort each include block alphabetically, but leave the blocks how you have them.

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    UINT32 exponent = (public->parameters).rsaDetail.exponent;
+
```
no newline here

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+    exponent = tpm2_util_hton_32(exponent);
+
+    ssl_rsa_key = RSA_new();
+
```
no newline here

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    fp = fopen(path, "wb");
+
```
no newline here and throughout where you check for a return value

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+    case pubkey_format_der:
+        ssl_res = i2d_RSA_PUBKEY_fp(fp, ssl_rsa_key);
+        break;
+    default:
```
Is this possible? Shouldn't this be a hard error as an unsupported format type?

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+        return files_save_bytes_to_file(path, (UINT8 *)signature,
+            sizeof(*signature));
+    }
+    else if (format == signature_format_plain) {
```
You have switches in some places for this and an if/else here, I would be consistent. I like the switch myself if we need to add another output format.

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    LOG_ERR("Unsupported signature output format.");
+    return false;
```
Yeah this is a hard error here, but not on line https://github.com/01org/tpm2-tools/pull/485/files#diff-98bdc2e9547136dd6f16812a7289b524R157

williamcroberts commented on lib/conversion.h:
```diff
@@ -0,0 +1,81 @@
+    signature_format_tss, signature_format_plain, signature_format_err
+};
+
+/*
```
doxygen comments here and throughout the header

williamcroberts commented on tools/tpm2_quote.c:
```diff
@@ -205,6 +210,80 @@ static UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
+        goto error;
+    }
+
+    num_bytes = calcSizeofTPM2B_ATTEST(quoted);
```
Where did this macro come from sapi headers?

Couldn't one use:
```
BUFFER_SIZE(typeof(quoted), field)
```

mgerstner replied:
I really don't want to touch this, because the `calcSizeofTPM2B_ATTEST()` uses the size field of the buffer while `BUFFER_SIZE()` uses `sizeof` the structure.

In my tests the data returned by Tss2_Sys_Quote() is much less than the size of the TPM2B_ATTEST structure for some reason. Either I don't understand some things here, or this might be worth opening an issue. However, changing to `BUFFER_SIZE()` would cause hundreds of zero bytes to be output into the quote file - at least on my machine.

The TPM chip actually only returns 121 bytes of data, while the structure is 1280 bytes large.

williamcroberts replied:
The statement, "I really don't want to touch this, because the calcSizeofTPM2B_ATTEST() uses the size field of the buffer while BUFFER_SIZE() uses sizeof the structure." is 100% wrong,

`BUFFER_SIZE()` uses the field within the struct.

williamcroberts replied:
I see your using the size stored in the structure, but why are you adding 2?

williamcroberts replied:
Ahh I see, now it makes sense it's not in the diffstat. Thos calcSize routines were inherited and already apart of tpm2_quote. Those routines look very wrong. I would remove those, as that's the cruft we want to get rid of in the repository.

We should be using the size fields and sizeof() and other non-magic ways to determine the sizes of buffers, or just serialize the buffer to disk.

The current problem of casting a struct to a byte() array, is that any padding goes to disk with it, and the file becomes machine dependent.

williamcroberts commented on tools/tpm2_quote.c:
```diff
@@ -205,6 +210,80 @@ static UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
+    FILE *fp = fopen(path,"wb");
+
+    if (!fp) {
+        printf("Output file %s cannot be created: %s\n",
```
LOG_ERR

williamcroberts replied:
And throughout

williamcroberts commented on tools/tpm2_quote.c:
```diff
@@ -205,6 +210,80 @@ static UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
+        goto error;
+    }
+
+    res = files_write_bytes(fp, (UINT8*)(quoted->b).buffer, (quoted->b).size);
```
Cant you replace this whole routine with ```files_save_bytes_to_file```

williamcroberts commented on tools/tpm2_readpublic.c:
```diff
@@ -92,8 +95,16 @@ static int read_public_and_save(TSS2_SYS_CONTEXT *sapi_context) {
 
-    return files_save_bytes_to_file(ctx.outFilePath, (UINT8 *) &public,
-            sizeof(public));
+    pubkey_format format = pubkey_format_tss;
```
set this in the static global, then you can just update it in the option handler, no need for checking if the f flag is set.

williamcroberts commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+bool tpm2_convert_signature(TPMT_SIGNATURE *signature, signature_format format, const char *path) {
+
+    if (format == signature_format_tss) {
+        /* TODO fix serialization */
```
We actually have been discussing serialization issues on: https://github.com/01org/tpm2-tss/pull/533

The issue here is that this dumps all padding, so its output not portable across architectures.

martinezjavier commented on lib/tpm2_alg_util.c:
```diff
@@ -330,3 +330,57 @@ bool pcr_parse_digest_list(char **argv, int len,
+
+    switch (signature->sigAlg) {
+    case TPM_ALG_RSASSA:
+        *size = sizeof((*signature).signature.rsassa.sig.t.buffer);
```
I would argue that using  the arrow operator is a much common idiom than doing a pointer dereference here, i.e:

```
*size = sizeof(signature->signature.rsassa.sig.t.buffer);
```
This comment is also for all the places in the code that do the same.

martinezjavier commented on lib/tpm2_alg_util.c:
```diff
@@ -330,3 +330,57 @@ bool pcr_parse_digest_list(char **argv, int len,
+        if(! buffer) {
+            goto nomem;
+        }
+        memcpy(buffer, &(hmac_sig->na), *size);
```
No need for the parenthesis here, the arrow operator has a higher precedence than the address-of operator.

martinezjavier commented on lib/tpm2_alg_util.c:
```diff
@@ -330,3 +330,57 @@ bool pcr_parse_digest_list(char **argv, int len,
+    case TPM_ALG_RSASSA:
+        *size = sizeof((*signature).signature.rsassa.sig.t.buffer);
+        buffer = malloc(*size);
+        if(! buffer) {
```
Please remove this space.

martinezjavier commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+//**********************************************************************;
+// Copyright (c) 2017, Intel Corporation
```
Shouldn't this be SUSE Linux GmbH instead?

martinezjavier commented on lib/conversion.c:
```diff
@@ -0,0 +1,204 @@
+bool tpm2_convert_pubkey(TPM2B_PUBLIC *public, pubkey_format format, const char *path) {
+
+    if (format == pubkey_format_der || format == pubkey_format_pem) {
+        return tpm2_convert_pubkey_ssl(&((public->t).publicArea), format, path);
```
Again, you can just write this as ```tpm2_convert_pubkey_ssl(&public->t.publicArea), format, path)``` since the ```->``` and ```.``` operators have more precedence than ```&```.

mgerstner commented:
I've adjusted according to your comments, what do you think now?

mgerstner commented:
I will rebase the changes for the final PR

williamcroberts commented:
@mgerstner it needs tests. We can deal with the sizeof calculation cleanups in a different PR, no need to do that here. The calc issue is #495

liuqun commented:
Hi @mgerstner 

There is a command line tool `pki --keyid` from project strongswan, which can be used to convert any tpm handle into a pubkey of OpenSSL PEM/DER format. Hope it‘ll be helpful for you.

- https://wiki.strongswan.org/projects/strongswan/wiki/TpmPlugin

mgerstner commented:
@williamcroberts what exactly needs tests? are there any changes for me left to do here then?

williamcroberts commented:
@mgerstner For testing, you could output to PEM format and make sure that openssl can use it.

mgerstner commented:
@williamcroberts Of course I've tested the changes and the output formats behave as (I) expected. I also ran the utilities through valgrind.

So are you requesting a unit test here?

williamcroberts commented:
@mgerstner no need for a unit test under test/unit. But integration tests under test/system would be great. We need something that exercises this code path in the test/system location.

liuqun commented:
There are some usefull tools to make OpenSSH work perfectly with TPM protected private keys

- libtpm2-pk11.so: PKCS11 provider for TPM 2.0 chips
https://github.com/irtimmer/tpm2-pk11

- libsimple-tpm-pk11.so: PKCS11 provider for TPM 1.2 chips
https://github.com/ThomasHabets/simple-tpm-pk11

mgerstner commented:
To get this PR into shape I've rebased the follow-up commits to clean up the history. I've sync-merged with your current master branch. I added man page documentation for the new switches.

Regarding the system test I considered just extending some of the existing tests like `test_tpm2_readpublic.sh`, however they mostly use the `TPM_ALG_KEYEDHASH` keys, so I can't test the OpenSSL logic with them.

So adding a dedicated test would be easier I guess. Do you run these tests against the tpm simulator?

I will be on vacation for the next two weeks so I can't work on this for a while.

williamcroberts commented:
@mgerstner we do run it against the simulator

mgerstner commented:
So here it is: a system test that covers the newly added functionality.

williamcroberts commented on man/tpm2_quote.1.md:
```diff
@@ -77,6 +84,12 @@ tpm2_quote -k 0x81010002 -P "hex:123abc" -L sha1:16,17,18+sha256:16,17,18 -o out
 
 [Github Issues](https://github.com/01org/tpm2-tools/issues)
 
+<<<<<<< HEAD
```
bad merge, ill fix this up.
